File: src/client/projeqtor-api.ts
import type { AppConfig } from "../config.js";
import { encryptAesCtr } from "./aes-ctr.js";
import { ProjeQtOrApiError } from "../utils/error-handler.js";
import type { Logger } from "../utils/logger.js";

export type ObjectClass = string;
export type WriteMethod = "PUT" | "POST" | "DELETE";
export interface SearchCriterion { field: string; operator?: string; value: string | number | boolean; }

export class ProjeQtOrApiClient {
  private readonly apiBase: string;
  private readonly authHeader: string;

  constructor(private readonly config: AppConfig, private readonly logger: Logger) {
    this.apiBase = `${config.PROJEQTOR_BASE_URL}/api`;
    this.authHeader = `Basic ${Buffer.from(`${config.PROJEQTOR_USERNAME}:${config.PROJEQTOR_PASSWORD}`).toString("base64")}`;
  }

  /** Retrieves an object by ProjeQtOr class and id. */
  getObject<T = unknown>(objectClass: ObjectClass, id: string | number): Promise<T> {
    return this.request<T>("GET", `/${encodeURIComponent(objectClass)}/${encodeURIComponent(String(id))}`);
  }

  /** Lists all objects of a ProjeQtOr class. */
  listAll<T = unknown>(objectClass: ObjectClass): Promise<T> {
    return this.request<T>("GET", `/${encodeURIComponent(objectClass)}/all`);
  }

  /** Retrieves objects using a saved ProjeQtOr filter id. */
  filter<T = unknown>(objectClass: ObjectClass, filterId: string | number): Promise<T> {
    return this.request<T>("GET", `/${encodeURIComponent(objectClass)}/filter/${encodeURIComponent(String(filterId))}`);
  }

  /** Searches objects. Criteria are converted to field:operator:value URL path segments. */
  search<T = unknown>(objectClass: ObjectClass, criteria: SearchCriterion[]): Promise<T> {
    const segments = criteria.map((c) => encodeURIComponent(`${c.field}:${c.operator ?? "= "}:${c.value}`.replace("= :", "=:")));
    return this.request<T>("GET", `/${encodeURIComponent(objectClass)}/search/${segments.join("/")}`);
  }

  /** Retrieves objects updated between ProjeQtOr timestamps YYYYMMDDHHMMSS. */
  updated<T = unknown>(objectClass: ObjectClass, from: string, to: string): Promise<T> {
    return this.request<T>("GET", `/${encodeURIComponent(objectClass)}/updated/${from}/${to}`);
  }

  /** Creates an object. Body is AES-CTR encrypted as required by ProjeQtOr writes. */
  create<T = unknown>(objectClass: ObjectClass, data: Record<string, unknown>): Promise<T> {
    return this.write<T>("PUT", objectClass, data);
  }

  /** Updates an object. Body is AES-CTR encrypted as required by ProjeQtOr writes. */
  update<T = unknown>(objectClass: ObjectClass, data: Record<string, unknown>): Promise<T> {
    return this.write<T>("POST", objectClass, data);
  }

  /** Deletes an object by sending an encrypted JSON payload, usually including id. */
  delete<T = unknown>(objectClass: ObjectClass, data: Record<string, unknown>): Promise<T> {
    return this.write<T>("DELETE", objectClass, data);
  }

  private write<T>(method: WriteMethod, objectClass: ObjectClass, data: Record<string, unknown>): Promise<T> {
    const json = JSON.stringify(data);
    const encrypted = encryptAesCtr(json, this.config.PROJEQTOR_API_KEY, this.config.PROJEQTOR_AES_KEY_LENGTH as 128 | 192 | 256);
    // Common ProjeQtOr API integrations expect an encrypted "data" field.
    return this.request<T>(method, `/${encodeURIComponent(objectClass)}`, { data: encrypted });
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.apiBase}${path}`;
    let attempt = 0;
    let lastError: unknown;
    while (attempt <= this.config.PROJEQTOR_RETRY_ATTEMPTS) {
      try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), this.config.PROJEQTOR_TIMEOUT_MS);
        const response = await fetch(url, {
          method,
          headers: {
            Authorization: this.authHeader,
            Accept: "application/json",
            ...(body === undefined ? {} : { "Content-Type": "application/json" })
          },
          body: body === undefined ? undefined : JSON.stringify(body),
          signal: controller.signal
        }).finally(() => clearTimeout(timeout));

        const text = await response.text();
        const parsed = text ? safeJson(text) : null;
        if (!response.ok) {
          const retryable = response.status === 429 || response.status >= 500;
          throw new ProjeQtOrApiError(extractApiMessage(parsed) ?? response.statusText, response.status, parsed, retryable);
        }
        return parsed as T;
      } catch (e) {
        lastError = e;
        const retryable = e instanceof ProjeQtOrApiError ? e.retryable : e instanceof Error && e.name === "AbortError";
        if (!retryable || attempt >= this.config.PROJEQTOR_RETRY_ATTEMPTS) break;
        const delay = this.config.PROJEQTOR_RETRY_BASE_DELAY_MS * Math.pow(2, attempt) + Math.floor(Math.random() * 100);
        this.logger.warn("Retrying ProjeQtOr API request", { method, path, attempt: attempt + 1, delay });
        await new Promise((resolve) => setTimeout(resolve, delay));
        attempt += 1;
      }
    }
    if (lastError instanceof ProjeQtOrApiError) throw lastError;
    if (lastError instanceof Error) throw new ProjeQtOrApiError(lastError.message, undefined, undefined, true);
    throw new ProjeQtOrApiError("Unknown API failure", undefined, undefined, true);
  }
}

function safeJson(text: string): unknown {
  try { return JSON.parse(text); } catch { return text; }
}

function extractApiMessage(value: unknown): string | undefined {
  if (value && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    for (const k of ["message", "error", "errorMessage", "result"]) {
      if (typeof obj[k] === "string") return obj[k] as string;
    }
  }
  return typeof value === "string" ? value : undefined;
}

File: projeqtor-mcp-server-python/src/projeqtor_mcp_server/client/projeqtor_api.py
from __future__ import annotations

import asyncio
import json
import logging
import random
from dataclasses import dataclass
from typing import Any, Literal
from urllib.parse import quote

import httpx

from projeqtor_mcp_server.client.aes_ctr import encrypt_aes_ctr
from projeqtor_mcp_server.config import Settings
from projeqtor_mcp_server.utils.errors import ProjeQtOrApiError

WriteMethod = Literal["PUT", "POST", "DELETE"]


@dataclass(frozen=True)
class SearchCriterion:
    """Criterion for ProjeQtOr `/search` endpoint."""

    field: str
    value: str | int | float | bool
    operator: str = "="

    def to_path_segment(self) -> str:
        return quote(f"{self.field}:{self.operator}:{self.value}", safe="")


class ProjeQtOrApiClient:
    """Async HTTP client wrapping the ProjeQtOr REST API."""

    def __init__(self, settings: Settings, logger: logging.Logger) -> None:
        self._settings = settings
        self._logger = logger
        self._api_base = f"{settings.base_url}/api"
        self._auth = httpx.BasicAuth(settings.projeqtor_username, settings.projeqtor_password)

    async def get_object(self, object_class: str, object_id: str | int) -> Any:
        """Retrieve one object by class and id."""
        return await self._request("GET", f"/{quote(object_class)}/{quote(str(object_id))}")

    async def list_all(self, object_class: str) -> Any:
        """List all objects of a class."""
        return await self._request("GET", f"/{quote(object_class)}/all")

    async def filter(self, object_class: str, filter_id: str | int) -> Any:
        """Retrieve objects using a predefined ProjeQtOr filter id."""
        return await self._request("GET", f"/{quote(object_class)}/filter/{quote(str(filter_id))}")

    async def search(self, object_class: str, criteria: list[SearchCriterion]) -> Any:
        """Search objects with SQL-like criteria encoded as URL path segments."""
        segments = "/".join(c.to_path_segment() for c in criteria)
        return await self._request("GET", f"/{quote(object_class)}/search/{segments}")

    async def updated(self, object_class: str, from_ts: str, to_ts: str) -> Any:
        """Retrieve objects updated between two ProjeQtOr timestamps."""
        return await self._request("GET", f"/{quote(object_class)}/updated/{from_ts}/{to_ts}")

    async def create(self, object_class: str, data: dict[str, Any]) -> Any:
        """Create an object with AES-CTR encrypted payload."""
        return await self._write("PUT", object_class, data)

    async def update(self, object_class: str, data: dict[str, Any]) -> Any:
        """Update an object with AES-CTR encrypted payload."""
        return await self._write("POST", object_class, data)

    async def delete(self, object_class: str, data: dict[str, Any]) -> Any:
        """Delete an object with AES-CTR encrypted payload."""
        return await self._write("DELETE", object_class, data)

    async def _write(self, method: WriteMethod, object_class: str, data: dict[str, Any]) -> Any:
        encrypted = encrypt_aes_ctr(
            json.dumps(data, ensure_ascii=False, separators=(",", ":")),
            self._settings.projeqtor_api_key,
            self._settings.projeqtor_aes_key_length,
        )
        return await self._request(method, f"/{quote(object_class)}", json_body={"data": encrypted})

    async def _request(self, method: str, path: str, json_body: Any | None = None) -> Any:
        url = f"{self._api_base}{path}"
        last_exc: BaseException | None = None
        timeout = httpx.Timeout(self._settings.projeqtor_timeout_seconds)

        for attempt in range(self._settings.projeqtor_retry_attempts + 1):
            try:
                async with httpx.AsyncClient(auth=self._auth, timeout=timeout) as http:
                    response = await http.request(
                        method,
                        url,
                        json=json_body,
                        headers={"Accept": "application/json"},
                    )
                payload = _decode_response(response)
                if response.is_error:
                    retryable = response.status_code == 429 or response.status_code >= 500
                    raise ProjeQtOrApiError(
                        _extract_message(payload) or response.reason_phrase,
                        status_code=response.status_code,
                        retryable=retryable,
                        details=payload,
                    )
                return payload
            except (httpx.TimeoutException, httpx.TransportError) as exc:
                last_exc = ProjeQtOrApiError(str(exc), retryable=True)
            except ProjeQtOrApiError as exc:
                last_exc = exc
                if not exc.retryable:
                    raise

            if attempt >= self._settings.projeqtor_retry_attempts:
                break
            delay = self._settings.projeqtor_retry_base_delay_seconds * (2**attempt) + random.random() / 10
            self._logger.warning("Retrying ProjeQtOr request", extra={"meta": {"method": method, "path": path, "attempt": attempt + 1, "delay": delay}})
            await asyncio.sleep(delay)

        if last_exc:
            raise last_exc
        raise ProjeQtOrApiError("Unknown ProjeQtOr API failure", retryable=True)


def _decode_response(response: httpx.Response) -> Any:
    text = response.text
    if not text:
        return None
    try:
        return response.json()
    except ValueError:
        return text


def _extract_message(payload: Any) -> str | None:
    if isinstance(payload, dict):
        for key in ("message", "error", "errorMessage", "result"):
            value = payload.get(key)
            if isinstance(value, str):
                return value
    return payload if isinstance(payload, str) else None
